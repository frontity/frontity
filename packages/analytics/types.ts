import { Package, Action } from "frontity/types";
import Source from "@frontity/source/types";
import Router from "@frontity/router/types";

/**
 * Pageview payload passed to analytics `pageview` actions.
 */
export interface Pageview {
  /**
   * Link of the page (without the host part).
   *
   * @example "/2016/the-beauties-of-gullfoss"
   */
  link: string;

  /**
   * Title of the page.
   *
   * @example "The Beauties Of Gullfoss - Frontity Test"
   */
  title: string;
}

/**
 * Eventy payload passed to analytics `event` actions.
 */
export interface Event {
  /**
   * The name of the event.
   *
   * @example "Click"
   */
  name: string;

  /**
   * Collection of properties sent along with the event.
   *
   * @remarks
   * Each analytics package requires some specific properties to be included
   * here. Check out their documentation for more info.
   *
   * @example
   * ```
   * {
   *   label: "video",
   *   category: "youtube",
   * }
   * ```
   */
  payload: Record<string, any>;
}

/**
 * Base types and actions to build analytics packages for Frontity.
 */
interface Analytics extends Package {
  /**
   * Root elements exposed by this package.
   */
  roots: {
    /**
     * Analytics root element.
     */
    analytics: React.FC;
  };

  /**
   * The actions exposed by this package.
   */
  actions: {
    /**
     * Analytics namespace.
     */
    analytics: {
      /**
       * Send a pageview to all active analytics packages.
       *
       * This action takes all namespaces defined in `state.analytics.pageviews`
       * that are `true` and calls the `pageview` action of each one with the
       * specified `Pageview` object.
       *
       * @remarks
       * This action is used by the `roots.analytics` component and is not
       * meant to be called directly.
       *
       * @example
       * ```
       * actions.analytics.pageview({
       *   link: "/2016/the-beauties-of-gullfoss",
       *   title: "The Beauties Of Gullfoss",
       * });
       * ```
       *
       * @param pageview - Object of type {@link Pageview}.
       */
      pageview: Action<Analytics, Pageview>;

      /**
       * Send an event to all active analytics packages.
       *
       * This action takes all namespaces defined in `state.analytics.events`
       * that are `true` and calls the `event` action of each one with the
       * specified `Event` object.
       *
       * @example
       * ```
       * actions.analytics.event({
       *   name: "click",
       *   payload: {
       *     category: "video",
       *     label: "featured-media",
       *   },
       * });
       * ```
       *
       * @param event - Object of type {@link Event}.
       */
      event: Action<Analytics, Event>;
    };
  };

  /**
   * The state exposed by this package.
   */
  state: {
    /**
     * Analytics namespace.
     */
    analytics: {
      /**
       * Map of namespaces with boolean values.
       *
       * This object is used by `actions.analytics.pageview` to know which
       * analytics packages should send the pageview to their respective
       * services.
       *
       * @remarks
       * If you want to disable sending pageviews for a specific analytics
       * package, the respective namespace of that package should be set here to
       * `false`.
       *
       * @example
       * ```
       * pageviews: {
       *   googleAnalytics: false,
       *   comscoreAnalytics: true
       * }
       * ```
       */
      pageviews: Record<string, boolean>;

      /**
       * Map of namespaces with boolean values.
       *
       * This object is used by `actions.analytics.event` to know which
       * analytics packages should send the event to their respective services.
       *
       * @remarks
       * If you want to disable sending events for a specific analytics
       * package, the respective namespace of that package should be set here to
       * `false`.
       *
       * @example
       * ```
       * events: {
       *   googleAnalytics: true,
       *   comscoreAnalytics: false,
       * }
       * ```
       */
      events: Record<string, boolean>;
    };
  };
}

/**
 * Package types used internally by Analytics.
 */
export type Packages = Analytics & Router & Source;

export default Analytics;
