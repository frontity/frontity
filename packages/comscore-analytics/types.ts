import { ReactType } from "react";
import { Action, Package } from "frontity/types";
import Analytics, { Pageview } from "@frontity/analytics/types";

/**
 * Comscore's Tag Parameters.
 */
interface ComscoreTagParams {
  /**
   * Tag Type. Pre-populated with a constant value of `2`.
   *
   * @defaultValue 2
   */
  c1?: string;

  /**
   * Unique ID for each client assigned by Comscore.
   */
  c2: string;

  /**
   * Full Page URL.
   *
   * Manually populated full page URL, including query string variables.
   */
  c4?: string;

  /**
   * Genre of content.
   *
   * Alphanumeric value used for client specific custom classification.
   */
  c5?: string;

  /**
   * Package.
   *
   * Alphanumeric value for customized aggregation to reflect sections or site
   * centric advertising packages.
   */
  c6?: string;

  /**
   * Client Segment ID.
   *
   * Alphanumeric value for Publisherâ€™s own segment for the machinethe content
   * asset is being served to.
   */
  c15?: string;
}
declare global {
  /**
   * Extended with Comscore global variables.
   */
  interface Window {
    /**
     * Comscore main object.
     */
    COMSCORE: {
      /**
       * Function that sends tag params to Comscore.
       *
       * @remarks
       * It generates the following params automatically, which are sent to
       * Comscore along with the received ones:
       *
       * - `c7`: Full Page URL, including query string variables.
       * - `c8`: Page Title.
       * - `c9`: Referring URL.
       *
       * @param tagParams - Object of type {@link ComscoreTagParams}.
       */
      beacon: (tagParams: ComscoreTagParams) => void;
    };

    /**
     * Array that stores the first tag params generated right before the
     * Comscore library has loaded.
     */
    _comscore: Array<ComscoreTagParams>;
  }
}

/**
 * Comscore Analytics package for Frontity.
 */
interface ComscoreAnalytics extends Package {
  /**
   * Root elements exposed by this package.
   */
  roots: Analytics["roots"] & {
    /**
     * Comscore Analytics root element.
     */
    comscoreAnalytics: ReactType;
  };

  /**
   * The state exposed by this package.
   */
  state: Analytics["state"] & {
    /**
     * Comscore Analytics namespace.
     */
    comscoreAnalytics: {
      /**
       * Single tracking ID associated to a Comscore client.
       */
      trackingId?: string;

      /**
       * Array of tracking IDs associated to a Comscore client.
       */
      trackingIds?: string[];
    };
  };

  /**
   * The actions exposed by this package.
   */
  actions: Analytics["actions"] & {
    /**
     * Comscore Analytics namespace.
     */
    comscoreAnalytics: {
      /**
       * Send a pageview to all Comscore traking ids defined in
       * `state.comscoreAnalytics.trackingIds`.
       *
       * @remarks
       * This action is called automatically if
       * `state.analytics.pageviews.comscoreAnalytics` is set to `true`.
       *
       * @param pageview - Object of type {@link Pageview}.
       */
      pageview: Action<ComscoreAnalytics, Pageview>;
    };
  };
}

/**
 * Package types used internally in ComscoreAnalytics.
 */
export type Packages = ComscoreAnalytics & Analytics;

export default ComscoreAnalytics;
