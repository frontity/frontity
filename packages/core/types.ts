import { TransformOptions } from "babel-core";
import { Configuration } from "webpack";

export type Mode = "development" | "production";
export type Target = "module" | "es5" | "server";

export interface BabelConfigs {
  module: TransformOptions;
  es5: TransformOptions;
  server: TransformOptions;
}

export interface WebpackConfigs {
  module: Configuration;
  es5: Configuration;
  server: Configuration;
}

export interface FrontityConfig {
  outDir: string;
}

export interface Config {
  babel: BabelConfigs;
  webpack: WebpackConfigs;
  frontity: FrontityConfig;
}

export interface EntryPoints {
  name: string;
  path: string;
}

export type FrontityTags = {
  script?: string;
  link?: string;
  style?: string;
};

export type HeadTags = {
  link: string;
  base: string;
  bodyAttributes: string;
  htmlAttributes: string;
  meta: string;
  noscript: string;
  script: string;
  style: string;
  title: string;
};

export type Template = ({
  html,
  frontity,
  head
}: {
  html: string;
  frontity: FrontityTags;
  head: HeadTags;
}) => string;
