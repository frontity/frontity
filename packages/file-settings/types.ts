type Base = {
  match?: string[];
  mode?: string;
  state?: object;
};

export type Package = {
  name: string;
  active?: boolean;
  state?: object;
};

export type NormalizedPackage = Package & {
  active: boolean; // Default: true
  state: object; // Default: {}
};

type Mono = Base & {
  name?: string;
};

type Multi = Base & {
  name: string;
};

type Imported<T> = Base & {
  packages: (string | T)[];
};

type Normalized<T> = Base & {
  mode: string; // Default: html
  state: object; // Default: {}
  packages: T[];
};

export type MonoSettings<T = Package> = Imported<T> & Mono;

export type MultiSettings<T = Package> = Imported<T> & Multi;

export type Settings<T = Package> = MonoSettings<T> | MultiSettings<T>[];

export type NormalizedSettings<T = NormalizedPackage> = Normalized<T> & Multi;

export type Site = {
  name: string;
  mode: string;
  packages: string[];
};
