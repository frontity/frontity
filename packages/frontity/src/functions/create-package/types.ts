// Options passed to the `create-package` function.
export type Options = {
  // Name of the package.
  name?: string;
  // Namespace of the package.
  namespace?: string;
  // Path of the Frontity project.
  projectPath?: string;
  // Path where the package should be created (relative to `projectPath`).
  packagePath?: string;
  // Support for TypeScript.
  typescript?: boolean;
};
