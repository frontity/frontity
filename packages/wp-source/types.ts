import { AsyncAction, Action, State, Derived } from "frontity/types";
import Source, { EntityData } from "@frontity/source/types";
import { Api } from "./src/libraries";

/**
 * A Frontity source package for the REST API of self-hosted WordPress and
 * WordPress.com sites.
 */
interface WpSource extends Source {
  /**
   * The name of this package.
   */
  name: "@frontity/wp-source";

  /**
   * The state exposed by this package.
   */
  state: {
    /**
     * Source namespace.
     *
     * It also contains the state defined in the main source package.
     */
    source: Source<WpSource>["state"]["source"] & {
      /**
       * The URL of the REST API.
       *
       * It can be from a self-hosted WordPress or from a WordPress.com site.
       *
       * @example "https://your-site.com/wp-json"
       * @example "https://public-api.wordpress.com/wp/v2/sites/your-site.wordpress.com"
       */
      api: string;

      /**
       * True when the REST API belongs to a WordPress.com site.
       */
      isWpCom: Derived<WpSource, boolean>;

      /**
       * The name or path indicating the subdirectory of the domain where the
       * Frontity site lives.
       *
       * For example, if your site is at https://mysite.com/blog, you have to
       * use it with the value of /blog.
       *
       * It is also used to transform links of the entities that come from the
       * REST API by prefixing them with this value.
       *
       * @example "/blog"
       */
      subdirectory: string;

      /**
       * Set a specific page as the homepage of the site.
       *
       * For example, if you set this value to `/about-us` then that page will
       * be shown when you access `/`.
       *
       * @remarks
       * - This option overrides the `/` route so it should be used in
       * combination with `state.source.postsPage` to be able to access the
       * posts archive with a different route.
       * - You will need to configure WordPress with the same setting.
       *
       * @example "/about-us"
       */
      homepage: string;

      /**
       * Show the posts archive when accessing a specific URL of your site,
       * instead of the homepage.
       *
       * For example, if you set this value to `/blog`, then the posts archive
       * will be shown if you access `/blog` instead of `/`.
       *
       * @remarks
       * - It is useful when used in combination with
       * `state.source.homepage`.
       * - You will need to configure WordPress with the same setting.
       *
       * @example "/blog"
       */
      postsPage: string;

      /**
       * Change the base prefix of the URLs for category pages.
       *
       * @remarks
       * Configure WordPress with the same setting.
       *
       * @example "categorias"
       */
      categoryBase: string;

      /**
       * Change the base prefix of the URLs for tag pages.
       *
       * @remarks
       * Configure WordPress with the same setting.
       *
       * @example "etiquetas"
       */
      tagBase: string;

      /**
       * Change the base prefix of the URLs for author pages.
       *
       * @remarks
       * Configure WordPress with the same setting.
       *
       * @example "autores"
       */
      authorBase: string;

      /**
       * Set the endpoint against which calls to the REST API are made when
       * posts are requested when fetching a single post, the post archive, date
       * archives, categories, tags, authors, and so on. This is useful when
       * you want to use another post type as your default.
       *
       * @example "products"
       *
       * @defaultValue "posts"
       */
      postEndpoint: string;

      /**
       * An object that will be used in each call to the REST API when
       * using `actions.source.fetch` with the default handlers.
       *
       * This is useful to filter fields from the REST API, change the default
       * `per_page` value and so on.
       *
       * @example
       * ```js
       * {
       *    per_page: 5,
       *    _fields: ["title", "id", "content", "link"]
       *  }
       * ```
       */
      params: Record<string, any>;

      /**
       * An array of objects that define the custom post types present in the
       * site.
       *
       * @example
       * ```js
       * [
       *   {
       *     type: "movie",
       *     endpoint: "movies",
       *     archive: "/movies-archive"
       *   }
       * ]
       * ```
       */
      postTypes: {
        /**
         * The slug of the custom post type as configured in WordPress.
         *
         * @example "movie"
         */
        type: string;

        /**
         * The REST API endpoint from where this post type can be fetched.
         *
         * @example "movies"
         */
        endpoint: string;

        /**
         * The URL of the archive for the custom post type.
         *
         * @example "/movies-archive"
         */
        archive?: string;
      }[];
      /**
       * An array of objects that define the custom taxonomies present in the
       * site.
       *
       * @example
       * ```js
       * [
       *    {
       *      taxonomy: "actors",
       *      endpoint: "actor",
       *      postTypeEndpoint: "movies",
       *      params: {
       *        per_page: 5,
       *        _embed: true
       *      }
       *    }
       * ]
       * ```
       */
      taxonomies: {
        /**
         * The slug of the custom taxonomy as configured in WordPress.
         *
         * @example "actors"
         */
        taxonomy: string;

        /**
         * The REST API endpoint from where this taxonomy can be fetched.
         *
         * @example "actor"
         */
        endpoint: string;

        /**
         * The REST API endpoint of the custom post type that should be
         * fetched for this taxonomy.
         *
         * @example "movies"
         */
        postTypeEndpoint?: string;

        /**
         * The params that should be used in the REST API when fetching this
         * taxonomy.
         *
         * @example
         * ```js
         * {
         *   per_page: 5,
         *   _embed: true
         * }
         * ```
         *
         * @defaultValue
         * ```js
         * {
         *   _embed: true
         * }
         * ```
         */
        params?: Record<string, any>;
      }[];
    };
  };

  /**
   * The actions exposed by this package.
   */
  actions: {
    /**
     * Source namespace.
     */
    source: {
      /**
       * An action that fetches all the information and entities related to a link.
       *
       * It populates the state with both:
       * - An entry in `state.source.data` with information about that link.
       * - Normalized entities in relevant part of the state, like
       * `state.source.post`, `state.source.category`
       * or `state.source.author` and so on.
       *
       * @param link - The link that should be fetched. It can be a URL or a
       * custom link created to fetch additional entities from the REST API.
       * - URLs start with `/`.
       * - Non URLs start with `@`.
       * @example `actions.source.fecth("/some-post");`
       * @example `actions.source.fecth("@comments/135");`
       *
       * @param options - Optional options.
       *
       * @returns A promise that resolves when the data fetching has finished.
       */
      fetch:
        | AsyncAction<WpSource, string>
        | AsyncAction<
            WpSource,
            string,
            {
              /**
               * Whether the fetch should be done again if data for that link
               * already exists.
               */
              force?: boolean;
            }
          >;

      /**
       * An internal action that bootstraps the initialization.
       *
       * @remarks
       * This action is not meant to be run by the user, but by the Frontity
       * framework.
       */
      init: Action<WpSource>;
    };
  };

  /**
   * The libraries exposed by this package.
   */
  libraries: {
    /**
     * Source namespace.
     */
    source: Source["libraries"]["source"] & {
      /**
       * A library helper to fetch the REST API.
       *
       * It has two methods, `api.init` and `api.get`:
       * - `api.init`: Used internally to initialize the class.
       * - `api.get`: Used to fetch the entities from the REST API.
       *
       * Types defined in {@link Api}.
       *
       * @example
       * `api.get({ endpoint: "pages", params: { _embed: true, include: '14' } });`
       * @example
       * `api.get({ endpoint: "posts", params: { _embed: true, categories: '2,3,4' } });`
       */
      api: Api;

      /**
       * A library helper to add entities to the Frontity state.
       *
       * @remarks
       * Entities are not overwritten. If an entity already exists in the state
       * and a new one is fetched, the one in the state will prevail. If you
       * want to overwrite them, use the `force` option.
       *
       * @example
       * ```js
       * const response = await libraries.source.api.get({ endpoint: "posts" });
       * const entities = await libraries.source.populate({ response, state });
       * ```
       */
      populate: (args: {
        /**
         * The Frontity state.
         */
        state: State<WpSource>;

        /**
         * The Response object.
         *
         * Usually returned from `api.get`, but can also be the one returned
         * by `window.fetch`.
         */
        response: Response;

        /**
         * The subdirectory of the Frontity site. When this options is passed,
         * this subdirectory is added to the entities' links.
         *
         * @example "/blog"
         *
         * @defaultValue `state.source.subdirectory`
         */
        subdirectory?: string;

        /**
         * A boolean that indicates if the entities should be overwritten.
         *
         * @defaultValue false
         */
        force?: boolean;
      }) => /**
       * Returns a promise that resolves with an array of objects with
       * attributes `type`, `id` and `link` representing the added entities.
       *
       * @example
       * ```js
       * const entities = await libraries.source.populate({ response, state });
       * data.entities = entities.map(entity => ({
       *  type: entity.type,
       *  id: entity.id,
       *  link: entity.link,
       * }))
       * ```
       */
      Promise<EntityData[]>;

      /**
       * Handlers are objects that associate a link pattern with a function
       * that fetches all the entities that belong to that WordPress link.
       *
       * Types defined in {@link Pattern}.
       */
      handlers: Pattern<Handler>[];

      /**
       * Redirections are objects that associate a link pattern with a function
       * that returns a new link.
       *
       * These redirections are used when `actions.source.fetch` is executed,
       * before the handlers.
       */
      redirections: Pattern<Redirection>[];

      /**
       * Extracts the total number of entities of an archive from its REST API
       * Response object.
       *
       * @param response - The Response object. Usually returned from
       * `api.get`, but can also be the one returned by `window.fetch`.
       * @param valueIfHeaderMissing - The value that must be returned if the
       * header containing the information is not found in the Response object.
       *
       * @returns - The total number of entities for that archive.
       *
       * @defaultValue 0
       */
      getTotal: (response: Response, valueIfHeaderMissing: number) => number;

      /**
       * Extracts the total number of pages of an archive from its REST API
       * Response object.
       *
       * @param response - The Response object. Usually returned from
       * `api.get`, but can also be the one returned by `window.fetch`.
       * @param valueIfHeaderMissing - The value that must be returned if the
       * header containing the information is not found in the Response object.
       *
       * @returns - The total number of pages for that archive.
       *
       * @defaultValue 0
       */
      getTotalPages: (
        response: Response,
        valueIfHeaderMissing: number
      ) => number;
    };
  };
}

export default WpSource;

/**
 * Handlers are objects that associate a link pattern with a function
 * that fetches all the entities that belong to that WordPress link.
 *
 * Handlers are used when `actions.source.fetch` is executed.
 */
export interface Pattern<F extends Function = Function> {
  /**
   * A unique name to identify this handler.
   *
   * @example "product"
   */
  name: string;

  /**
   * The priority of this handler.
   *
   * Lower numbers have higher priority.
   *
   * @defaultValue 10
   */
  priority: number;

  /**
   * The pattern used to compare against the link that is being fetched.
   *
   * It uses [`path-to-regexp`](https://github.com/pillarjs/path-to-regexp)
   * under the hood, so please check its documentation to know how to write
   * patterns.
   *
   * @example "/product/:slug"
   */
  pattern: string;

  /**
   * The function that does the actual fetching.
   *
   * Types defined in {@link Handler}.
   */
  func: F;
}

/**
 * The handler function is in charge of fetching all the entities that belong
 * to a link.
 *
 * For example, when we fetch the link of a page, the handler must retrieve the
 * entitiy for that page, but also its author. When we fetch the link of a post,
 * the handler should retrieve the entity of the post, its author, its
 * categories, its tags, its featured image and so on.
 *
 * @example
 * ```js
 * libraries.source.handlers.push({
 *    name: "product",
 *    priority: 10,
 *    pattern: "/product/:slug",
 *    func: async ({ link, params, state, libraries, force }) => {
 *      // 1. Get the product entities from the REST API.
 *      const response = await libraries.source.api.get({
 *        endpoint: "products",
 *        params: { slug: params.slug }
 *      });
 *
 *      // 2. Add the product to the state.
 *      const [product] = await libraries.source.populate({ response, state, force });
 *
 *      // 3. Add the data object for this link.
 *      Object.assign(state.source.data[link], {
 *        id: product.id,
 *        type: product.type,
 *        isPostType: true,
 *        isProduct: true
 *      });
 *    }
 *  });
 * ```
 *
 * @param args - The arguments object.
 *
 * @returns - Returns a promise that is resolved when the handler has finished.
 */
export interface Handler<Pkg extends Source = WpSource> {
  (args: {
    /**
     * The link that is being fetched.
     *
     * @example "/some-post"
     */
    link: string;

    /**
     * The link that is being fetched.
     *
     * @deprecated
     * Use `link` instead.
     */
    route: string;

    /**
     * The values extracted from the pattern.
     *
     * For example, if the pattern is `"/category/:slug"` and it is called for
     * the link `"/category/nature"`, it will receive `{ slug: "nature" }`.
     *
     * @example
     * ```js
     * {
     *  slug: "nature"
     * }
     * ```
     */
    params: { [param: string]: any };

    /**
     * The Frontity state.
     */
    state: State<Pkg>;

    /**
     * The Frontity libraries.
     */
    libraries: Pkg["libraries"];

    /**
     * Whether `actions.source.fetch` was called with the `force` option.
     */
    force?: boolean;
  }): Promise<void>;
}

/**
 * Redirections are objects that associate a link pattern with a function
 * that returns a new link.
 *
 * These redirections are used when `actions.source.fetch` is executed,
 * before the handlers.
 *
 * @example
 * ```js
 * libraries.source.redirections.push({
 *   name: "tags-to-labels",
 *   priority: 5,
 *   pattern: "/tag/:slug/",
 *   func: ({ slug }) => "/label/" + slug
 * });
 * ```
 *
 * @param params - An object that contains the values extracted from the
 * pattern.
 *
 * For example, if the pattern is `"/category/:slug"` and it is called for
 * the link `"/category/nature"`, it will receive `{ slug: "nature" }`.
 *
 * @example
 * ```js
 * {
 *  slug: "nature"
 * }
 * ```
 *
 * @returns The new link.
 */
export interface Redirection {
  (params?: Record<string, string>): string;
}
