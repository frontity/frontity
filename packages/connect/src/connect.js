import {
  useState,
  useEffect,
  useMemo,
  memo,
  useRef,
  createContext,
  useContext,
} from "react";
import { warn } from "@frontity/error";
import { observe, unobserve } from "@frontity/observer-util";

const COMPONENT = Symbol("owner component");

/**
 * A React context that stores the store of Frontity Connect.
 */
const context = createContext({});

/**
 * The Provider of the store context, to be used in the root of the React
 * application.
 */
export const Provider = context.Provider;

/**
 * A flag to know when a connected component is being rendered.
 */
let isConnected = false;

/**
 * The default options of connect.
 */
const defaultOptions = {
  /**
   * Define if `connect` should inject the store props in the component, or only
   * make it reactive. Useful when the store is accessed via `useConnect` and
   * the props are passed down an HTML node.
   */
  injectProps: true,
};

/**
 * Flag to know if there's been a change before the component was mounted.
 */
const changedBeforeMounted = new WeakMap();

/**
 * Flag to know when a component has been mounted.
 */
const mounted = new WeakMap();

/**
 * Connect a React component with the Frontity connect store and make it
 * reactive to its changes.
 *
 * @param Comp - The React component to be connected.
 * @param options - The connect options, defined in {@link defaultOptions}.
 *
 * @returns The same Component, but reactive and with the store included in the
 * props.
 */
export function connect(Comp, options) {
  const isFunctionalComp = !(Comp.prototype && Comp.prototype.isReactComponent);
  options = options ? { ...defaultOptions, ...options } : defaultOptions;

  let ReactiveComp;

  if (isFunctionalComp) {
    // Use a hook based reactive wrapper.
    ReactiveComp = (props) => {
      // Dummy setState to update the component.
      const [, setState] = useState();

      // Use a ref to store the reaction because we want to do some annotations.
      const reaction = useRef();

      // Create a memoized reactive wrapper of the original component (render)
      // at the very first run of the component function.
      const render = useMemo(
        () => {
          reaction.current = observe(Comp, {
            scheduler: () => {
              // Trigger a new rerender if the component has already been
              // mounted.
              if (mounted.get(reaction.current)) setState({});
              // Annotate it as "changed" if the component has not been mounted
              // yet.
              else changedBeforeMounted.set(reaction.current, true);
            },
            lazy: true,
          });

          // Initialize a flag to know if the component was finally mounted.
          mounted.set(reaction.current, false);

          // Initialize a flag to know if the was reaction was invalidated
          // before the component was mounted.
          changedBeforeMounted.set(reaction.current, false);

          return reaction.current;
        },

        // Adding the original Comp here is necessary to make HMR work. It does
        // not affect behavior otherwise.
        // eslint-disable-next-line react-hooks/exhaustive-deps
        [Comp]
      );

      useEffect(() => {
        // Mark the component as mounted.
        mounted.set(reaction.current, true);

        // If there was a change before the component was mounted, trigger a
        // new rerender.
        if (changedBeforeMounted.get(reaction.current)) setState({});

        // Cleanup the reactive connections when the component is unmounted.
        return () => unobserve(reaction.current);
      }, []);

      // Extract the Frontity store from the context.
      const { state, actions, libraries } = useContext(context);

      // The isConnected flag is used to warn users that try to use `useConnect`
      // in a non-connected component.
      isConnected = true;
      try {
        // Run the reactive render instead of the original one and inject the
        // props if necessary.
        return render({
          ...props,
          ...(options.injectProps && state && { state }),
          ...(options.injectProps && actions && { actions }),
          ...(options.injectProps && libraries && { libraries }),
        });
      } finally {
        isConnected = false;
      }
    };
  } else {
    /**
     * Wrap the component to extract the store from the context and pass it
     * down to the final component. Also, make the render method reactive.
     */
    class ReactiveClassComp extends Comp {
      /**
       * Initialize the wrapper.
       *
       * @param props - The props passed to the original component.
       * @param context - The React context.
       */
      constructor(props, context) {
        super(props, context);

        // Create some state to be able to rerender the component later.
        this.state = this.state || {};
        this.state[COMPONENT] = this;

        // Make the render method reactive.
        this.render = observe(this.render, {
          scheduler: () => this.setState({}),
          lazy: true,
        });
      }

      /**
       * The render method of this React Class.
       *
       * @returns The same Component but with the store injected as props.
       */
      render() {
        const { state, actions, libraries } = this.context;
        const props = { ...this.props, state, actions, libraries };
        return <Comp {...props} />;
      }

      /**
       * React should trigger updates on prop changes, while  handles
       * store changes.
       *
       * @param nextProps - The props of the next render.
       * @param nextState - The state of the next render.
       * @returns Whether the component should rerender or not.
       */
      shouldComponentUpdate(nextProps, nextState) {
        const { props, state } = this;

        // Respect the case when the user defines a shouldComponentUpdate.
        if (super.shouldComponentUpdate) {
          return super.shouldComponentUpdate(nextProps, nextState);
        }

        // Return true if it is a reactive render or the state changes.
        if (state !== nextState) {
          return true;
        }

        // The component should also update if any of its props shallowly
        // changed value.
        const keys = Object.keys(props);
        const nextKeys = Object.keys(nextProps);
        return (
          nextKeys.length !== keys.length ||
          nextKeys.some((key) => props[key] !== nextProps[key])
        );
      }

      /**
       * Remove the reaction when the component is unmounted.
       */
      componentWillUnmount() {
        // Call user defined componentWillUnmount.
        if (super.componentWillUnmount) {
          super.componentWillUnmount();
        }
        // Clean up memory used.
        unobserve(this.render);
      }
    }

    // Attach the context and pass the compnent up.
    ReactiveClassComp.contextType = context;
    ReactiveComp = ReactiveClassComp;
  }

  // Make sure we display the correct name.
  ReactiveComp.displayName = Comp.displayName || Comp.name;

  // Static props are inherited by class components, but have to be copied for
  // function components.
  if (isFunctionalComp) {
    Object.keys(Comp).forEach((key) => {
      ReactiveComp[key] = Comp[key];
    });
  }

  return isFunctionalComp ? memo(ReactiveComp) : ReactiveComp;
}

/**
 * React hook that returns the Frontity store in connected components.
 *
 * @returns The Frontity store (state, actions, libraries...).
 */
export const useConnect = () => {
  if (!isConnected)
    warn(
      "Warning: useConnect() is being used in a non connected component, " +
        "therefore the component won't update on state changes. " +
        "Please wrap your component with connect().\n"
    );

  return useContext(context);
};
