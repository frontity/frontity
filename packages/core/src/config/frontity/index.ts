import { FrontityConfig } from "../../../types";

export default (): FrontityConfig => ({
  outDir: "build"
});
