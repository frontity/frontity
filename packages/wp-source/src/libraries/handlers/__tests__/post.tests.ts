import { State } from "frontity/types";
import { createStore } from "@frontity/connect";
import WpSource from "../../../../types";
import handler from "../post";
import { mockResponse } from "./mocks/helpers";

import wpSource from "../../../";
jest.mock("../../../");

const post60 = {
  id: 60,
  date: "2016-11-25T18:31:11",
  slug: "the-beauties-of-gullfoss",
  type: "post",
  link: "https://test.frontity.io/2016/the-beauties-of-gullfoss/",
  author: 4,
  featured_media: 62,
  meta: [],
  categories: [57, 59, 56, 3, 8, 58],
  tags: [10, 9, 13, 11]
};

let state: State<WpSource>;
let libraries: WpSource["libraries"];

beforeEach(() => {
  const config = wpSource();
  const { api, populate } = config.libraries.source;

  // mock api
  (api.get as jest.Mock).mockResolvedValue(mockResponse([post60]));

  // mock populate
  (populate as jest.Mock).mockResolvedValue([
    {
      id: 60,
      slug: "the-beauties-of-gullfoss",
      link: "https://test.frontity.io/2016/the-beauties-of-gullfoss/"
    }
  ]);

  ({ state, libraries } = createStore(config));
});

describe("post", () => {
  test("doesn't exist in source.post", async () => {
    // source.fetch("/the-beauties-of-gullfoss/")
    state.source.data["/the-beauties-of-gullfoss/"] = {
      isFetching: true,
      isReady: false
    };

    await handler({
      route: "/the-beauties-of-gullfoss/",
      params: { slug: "the-beauties-of-gullfoss" },
      state,
      libraries
    });

    expect(state.source.data).toMatchSnapshot();
  });

  test("exists in source.post", async () => {
    const get = libraries.source.api.get as jest.Mock;

    state.source.post[60] = post60;

    // source.fetch("/the-beauties-of-gullfoss/")
    state.source.data["/the-beauties-of-gullfoss/"] = {
      isFetching: true,
      isReady: false
    };

    await handler({
      route: "/the-beauties-of-gullfoss/",
      params: { slug: "the-beauties-of-gullfoss" },
      state,
      libraries
    });

    expect(get).not.toBeCalled();
    expect(state.source.data).toMatchSnapshot();
  });

  test("throws an error if it doesn't exist", async () => {
    libraries.source.api.get = jest
      .fn()
      .mockResolvedValueOnce(mockResponse([]));

    libraries.source.populate = jest
      .fn()
      .mockResolvedValueOnce(mockResponse([]));

    // source.fetch("/the-beauties-of-gullfoss/")
    state.source.data["/the-beauties-of-gullfoss/"] = {
      isFetching: true,
      isReady: false
    };

    const promise = handler({
      route: "/the-beauties-of-gullfoss/",
      params: { slug: "the-beauties-of-gullfoss" },
      state,
      libraries
    });

    expect(promise).rejects.toThrowError();
  });
});
