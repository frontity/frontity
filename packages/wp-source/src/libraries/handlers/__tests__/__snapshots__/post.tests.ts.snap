// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`post doesn't exist in source.post 1`] = `
Object {
  "/the-beauties-of-gullfoss/": Object {
    "id": 60,
    "isFetching": true,
    "isPost": true,
    "isPostType": true,
    "isReady": false,
    "type": "post",
  },
}
`;

exports[`post exists in source.post 1`] = `
Object {
  "/the-beauties-of-gullfoss/": Object {
    "id": 60,
    "isFetching": true,
    "isPost": true,
    "isPostType": true,
    "isReady": false,
    "type": "post",
  },
}
`;
